department-of-veterans-affairs/caseflow
# Message user about first job failure after 24 hours

Connects #12689 

### Description
User gets an Inbox message when any of their jobs encounters a failure after 24 hours of trying.

### Acceptance Criteria
- [ ] When a job has been re-tried more than 24 hours after it has been submitted, and it still fails, send a message to the users inbox
- [ ] In order to prevent duplicates, do not send this message to the user if they have already received one

### Testing Plan
Use a feature spec that already creates a fake job with appropriately old timestamps, e.g.
1. Instrument and run `spec/feature/asyncable_jobs/asyncable_jobs_index_spec.rb:61`
2. `pry> hlr.update_error!("oops!")`
3. Switch to the user mentioned in the resulting `Message` object
4. Visit `/inbox` and verify that message landed in the inbox 
5. Also added a group of tests for the `AsyncableJobMessaging` service class, for all the edge cases

### User Facing Changes
<img width="909" alt="Screen Shot 2019-11-20 at 13 57 38" src="https://user-images.githubusercontent.com/282869/69270823-c1f63600-0ba1-11ea-8b44-2e2a218896e4.png">

### Notes
- Is there a way to fully simulate the processing of a fake job, in the tests?
- The messaging logic is currently triggered by `Asyncable#update_error!` which feels a bit tightly coupled to the storage side of the ORM. However, any alternative would require each Asyncable implementation to explicitly trigger messaging, which is brittle and non-DRY. Thoughts?
